std/selector/slice.go

```diff
@@ -0,0 +1,87 @@
+package selector
+
+import (
+	"github.com/consensys/gnark/constraint/solver"
+	"github.com/consensys/gnark/frontend"
+	"math/big"
+)
+
+func init() {
+	// register hints
+	solver.RegisterHint(StepOutput)
+}
+
+// Partition selects left or right side of the input array, with respect to the pivotPosition.
+// More precisely when rightSide is false, for each i we have:
+//
+//	if i < pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// and when rightSide is true, for each i we have:
+//
+//	if i >= pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// We must have pivotPosition >= 1 and pivotPosition <= len(input)-1, otherwise a proof cannot be generated.
+func Partition(api frontend.API, pivotPosition frontend.Variable, rightSide bool,
+	input []frontend.Variable) (out []frontend.Variable) {
+	out = make([]frontend.Variable, len(input))
+	var mask []frontend.Variable
+	// we create a bit mask to multiply with the input
+	if rightSide {
+		mask = StepMask(api, len(input), pivotPosition, 0, 1)
+	} else {
+		mask = StepMask(api, len(input), pivotPosition, 1, 0)
+	}
+	for i := 0; i < len(out); i++ {
+		out[i] = api.Mul(mask[i], input[i])
+	}
+	return
+}
+
+// StepMask generates a step like function into an output array of a given length.
+// The output is an array of length outputLen,
+// such that its first stepPosition elements are equal to startValue and the remaining elements are equal to
+// endValue. Note that outputLen cannot be a circuit variable.
+//
+// We must have stepPosition >= 1 and stepPosition <= outputLen, otherwise a proof cannot be generated.
+// This function panics when outputLen is less than 2.
+func StepMask(api frontend.API, outputLen int,
```

Comment:
I get the function and its use. However I am not sure if it makes sense to have it public. It took me a minute to understand what it does (from method doc). Is there any potential use outside of `Partition` function?

Reply:
Actually, I think we should implement a `Slice` method. The thing is, after a while I realized that cascading is not a good idea and working with `mask` and using addition and subtraction can be more efficient. That's why I made `StepMask` public (exported). For example:
```
mask1         = [0, 0, 0, 1, 1, 1, 1, 1]
mask2         = [0, 0, 0, 0, 0, 0, 1, 1]
mask1 - mask2 = [0, 0, 0, 1, 1, 1, 0, 0]
```
So we will avoid an array multiplication.

I think implementing `SliceCut` won't be so easy, since `end` and `start` are wires. We would need to use `Mux` to *shift* every element.

Reply:
Now I'm not sure do we really need a `Slice` function? Cascading seems pretty straightforward.

Reply:
Actually, now looking at it, I think `Slice` function makes a bit more sense that `Partition`. I really do not like the `rightSide` argument in the argument of `Partition`. I think it is error-prone for the users, while `Slice` is really natural when working with slices (except that the returned array has the same length).

Reply:
So maybe we should make `Partiion` unexported and replace it with a `Slice` function?

Reply:
ops, that doesn't work. It will make one sided selection inefficient. 
What about having two functions `RightPartition` and `LeftPartition`?

Reply:
There is some special limitations because we are working with a circuit.

Reply:
In that case, I think lets keep it as is.

Reply:
OK then, Can you please push your changes for hints?

Reply:
I'll also add a `Slice` function it could be removed later if it was not useful.

Reply:
> OK then, Can you please push your changes for hints?

Will do